shell/platform/fuchsia/flutter/software_surface.cc

```diff
@@ -0,0 +1,404 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "software_surface.h"
+
+#include <lib/async/default.h>
+#include <lib/ui/scenic/cpp/commands.h>
+#include <zircon/status.h>
+#include <zircon/types.h>
+
+#include <algorithm>
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/trace_event.h"
+#include "include/core/SkImageInfo.h"
+#include "third_party/skia/include/core/SkColorSpace.h"
+#include "third_party/skia/include/core/SkImageInfo.h"
+
+#include "../runtime/dart/utils/inlines.h"
+
+namespace flutter_runner {
+
+namespace {
+
+constexpr SkColorType kSkiaColorType = kRGBA_8888_SkColorType;
+
+}  // namespace
+
+SoftwareSurface::SoftwareSurface(
+    fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+    fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+    scenic::Session* session,
+    const SkISize& size,
+    uint32_t buffer_id)
+    : session_(session), wait_(this) {
+  FML_CHECK(session_ || flatland_allocator.is_bound());
+
+  if (!SetupSkiaSurface(sysmem_allocator, flatland_allocator, size,
+                        buffer_id)) {
+    FML_LOG(ERROR) << "SoftwareSurface: Could not create render surface.";
+    return;
+  }
+
+  if (!CreateFences()) {
+    FML_LOG(ERROR) << "SoftwareSurface: Could not create signal fences.";
+    return;
+  }
+
+  PushSessionImageSetupOps(session);
+
+  wait_.set_object(release_event_.get());
+  wait_.set_trigger(ZX_EVENT_SIGNALED);
+  Reset();
+
+  valid_ = true;
+}
+
+SoftwareSurface::~SoftwareSurface() {
+  if (session_) {
+    if (image_id_) {
+      session_->Enqueue(scenic::NewReleaseResourceCmd(image_id_));
+    }
+    if (buffer_id_) {
+      session_->DeregisterBufferCollection(buffer_id_);
+    }
+  } else {
+    release_image_callback_();
+  }
+  wait_.Cancel();
+  wait_.set_object(ZX_HANDLE_INVALID);
+}
+
+bool SoftwareSurface::IsValid() const {
+  return valid_;
+}
+
+SkISize SoftwareSurface::GetSize() const {
+  if (!valid_) {
+    return SkISize::Make(0, 0);
+  }
+
+  return SkISize::Make(sk_surface_->width(), sk_surface_->height());
+}
+
+bool SoftwareSurface::CreateFences() {
+  if (zx::event::create(0, &acquire_event_) != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to create acquire event";
+    return false;
+  }
+
+  if (zx::event::create(0, &release_event_) != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to create release event";
+    return false;
+  }
+
+  return true;
+}
+
+bool SoftwareSurface::SetupSkiaSurface(
+    fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+    fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+    const SkISize& size,
+    uint32_t buffer_id) {
+  if (size.isEmpty()) {
+    FML_LOG(ERROR)
+        << "SoftwareSurface: Failed to allocate surface, size is empty";
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionTokenSyncPtr local_token;
```

Comment:
Would it be useful to leave a comment here explaining the roles of:

- local_token
- scenic_token
- import_token
- export_token?

It's hard for me to tell the differences between them.

Reply:
Done